### PR TITLE
feat: implement retry loop for Commit()

### LINKED ISSUE #648: Add retry loop for internal::ConnectionImpl::Commit
We need to:

- [x] Implement the functionality.
- [x] Write a unit test verifying that it tolerates at least one transient error.
- [x] Write a unit test verifying that it stops on the first permanent error.
- [x] Write a unit test verifying that it stops when there are too many transient failures.

See the discussion on #627 this may require changes to the implementation too.

### PR DESCRIPTION
**Pull request by coryan**
Fixes #648

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/680)

### CODE REVIEW

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 1]
```diff
@@ -626,6 +652,78 @@ TEST(ConnectionImplTest, CommitCommitFailure) {
...
+                          spanner_proto::CommitRequest const& request) {
+        EXPECT_EQ("test-session-name", request.session());
+        EXPECT_EQ("test-txn-id", request.transaction_id());
+        spanner_proto::CommitResponse response;
+        *response.mutable_commit_timestamp() =
+            internal::ToProto(Timestamp{std::chrono::seconds(123)});
+        return response;
+      }));
+
+  auto txn = MakeReadWriteTransaction();
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl_test.cc]: Should we add a comment, like below, about why we're binding an id?

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 2]
```diff
@@ -626,6 +652,78 @@ TEST(ConnectionImplTest, CommitCommitFailure) {
...
+  internal::Visit(txn, [](SessionHolder&, spanner_proto::TransactionSelector& s,
+                          std::int64_t) {
+    s.set_id("test-txn-id");
+    return 0;
+  });
+
+  auto commit = conn->Commit({txn, {}});
+  EXPECT_STATUS_OK(commit);
```
Comment (devbww) [inline google/cloud/spanner/internal/connection_impl_test.cc]: EXPECT_EQ(commit->commit_timestamp, Timestamp{std::chrono::seconds(123)}); ?

File: google/cloud/spanner/internal/connection_impl_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/spanner/internal/connection_impl_test.cc]: Done.

File: google/cloud/spanner/internal/connection_impl_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/spanner/internal/connection_impl_test.cc]: Done.

**Pull request merged**